File: backend/app/domains/admin/schemas/failed_jobs.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

FailedJobStatus = Literal["failed", "retrying", "resolved", "cancelled"]
FailedJobAction = Literal["retry", "cancel", "mark_resolved"]

JOB_TYPE_LABELS: dict[str, str] = {
    "documents.process": "extraction",
    "documents.delete": "deletion_cleanup",
    "documents.reindex": "reindex",
    "evaluations.run": "evaluation",
}


class FailedJobSummary(BaseModel):
    id: UUID
    organization_id: UUID
    task_id: str
    task_name: str
    job_type: str
    status: str
    queue_name: str | None
    error_code: str | None
    attempt_count: int
    is_retryable: bool
    entity_type: str | None
    entity_id: UUID | None
    last_attempted_at: datetime | None
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime


class FailedJobAuditEntry(BaseModel):
    id: UUID
    action: str
    performed_by_id: UUID | None
    note: str | None
    created_at: datetime


class FailedJobDetail(FailedJobSummary):
    error_message: str | None
    metadata_json: dict = Field(default_factory=dict)
    audit_log: list[FailedJobAuditEntry] = Field(default_factory=list)


class FailedJobsListResponse(BaseModel):
    items: list[FailedJobSummary]
    total: int
    page: int
    page_size: int


class BulkRetryRequest(BaseModel):
    job_ids: list[UUID] = Field(..., min_length=1, max_length=100)


class BulkRetryResponse(BaseModel):
    queued: list[UUID]
    skipped: list[UUID]
    skip_reasons: dict[str, str]

File: frontend/src/lib/api/failed-jobs.ts
import { apiRequest } from "@/lib/api/request";

export type FailedJobStatus = "failed" | "retrying" | "resolved" | "cancelled";

export type FailedJobSummary = {
  id: string;
  organization_id: string;
  task_id: string;
  task_name: string;
  job_type: string;
  status: FailedJobStatus;
  queue_name: string | null;
  error_code: string | null;
  attempt_count: number;
  is_retryable: boolean;
  entity_type: string | null;
  entity_id: string | null;
  last_attempted_at: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
};

export type FailedJobAuditEntry = {
  id: string;
  action: string;
  performed_by_id: string | null;
  note: string | null;
  created_at: string;
};

export type FailedJobDetail = FailedJobSummary & {
  error_message: string | null;
  metadata_json: Record<string, unknown>;
  audit_log: FailedJobAuditEntry[];
};

export type FailedJobsListResponse = {
  items: FailedJobSummary[];
  total: number;
  page: number;
  page_size: number;
};

export type BulkRetryResponse = {
  queued: string[];
  skipped: string[];
  skip_reasons: Record<string, string>;
};

export type FailedJobsQuery = {
  job_type?: string;
  status?: FailedJobStatus;
  queue_name?: string;
  retryable_only?: boolean;
  page?: number;
  page_size?: number;
};

export async function listFailedJobs(
  query?: FailedJobsQuery,
): Promise<FailedJobsListResponse> {
  const params = new URLSearchParams();
  if (query?.job_type) params.set("job_type", query.job_type);
  if (query?.status) params.set("status", query.status);
  if (query?.queue_name) params.set("queue_name", query.queue_name);
  if (query?.retryable_only) params.set("retryable_only", "true");
  if (query?.page != null) params.set("page", String(query.page));
  if (query?.page_size != null) params.set("page_size", String(query.page_size));
  const qs = params.toString();
  return apiRequest<FailedJobsListResponse>(
    `/admin/failed-jobs${qs ? `?${qs}` : ""}`,
  );
}

export async function getFailedJob(jobId: string): Promise<FailedJobDetail> {
  return apiRequest<FailedJobDetail>(`/admin/failed-jobs/${jobId}`);
}

export async function retryFailedJob(
  jobId: string,
): Promise<FailedJobSummary> {
  return apiRequest<FailedJobSummary>(`/admin/failed-jobs/${jobId}/retry`, {
    method: "POST",
  });
}

export async function bulkRetryFailedJobs(
  jobIds: string[],
): Promise<BulkRetryResponse> {
  return apiRequest<BulkRetryResponse>(`/admin/failed-jobs/bulk-retry`, {
    method: "POST",
    json: { job_ids: jobIds },
  });
}

export async function cancelFailedJob(
  jobId: string,
): Promise<FailedJobSummary> {
  return apiRequest<FailedJobSummary>(`/admin/failed-jobs/${jobId}/cancel`, {
    method: "POST",
  });
}

export async function resolveFailedJob(
  jobId: string,
): Promise<FailedJobSummary> {
  return apiRequest<FailedJobSummary>(`/admin/failed-jobs/${jobId}/resolve`, {
    method: "POST",
  });
}
